{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from matplotlib_venn import venn3\n",
    "import matplotlib.pyplot as plt\n",
    "import os\n",
    "import pyranges as pr\n",
    "import gseapy as gp\n",
    "from gseapy import barplot, dotplot\n",
    "import json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/che/TRIM/git/tcr/figures/saliency/include_emergent_disappear/CD8/data/compiled\n",
      "/home/che/TRIM/git/tcr/figures/saliency/include_emergent_disappear/CD8/data/compiled/venn/all_gradient\n"
     ]
    }
   ],
   "source": [
    "cell_type = 'CD8'\n",
    "de_type = 'include_emergent_disappear' # include_emergent_disappear, no_emergent_include_disappear, no_emergent_no_disappear\n",
    "fig_path = '/home/che/TRIM/git/tcr/figures/saliency'\n",
    "parent_path = f'{fig_path}/{de_type}/{cell_type}/data'\n",
    "geneset_path = f'{parent_path}/gene_sets'\n",
    "csv_path = f'{parent_path}/diffexp_df.csv'\n",
    "output_path = f'{parent_path}/compiled'\n",
    "venn_path = f'{output_path}/venn/all_gradient'\n",
    "\n",
    "if not os.path.exists(output_path):\n",
    "    os.makedirs(output_path)\n",
    "if not os.path.exists(venn_path):\n",
    "    os.makedirs(venn_path)\n",
    "\n",
    "print(output_path)\n",
    "print(venn_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>lfc_post</th>\n",
       "      <th>lfc_pre</th>\n",
       "      <th>lfc_clone_pre_10</th>\n",
       "      <th>lfc_clone_pre_1</th>\n",
       "      <th>lfc_post_pvals_adj</th>\n",
       "      <th>lfc_pre_pvals_adj</th>\n",
       "      <th>lfc_clone_pre_10_pvals_adj</th>\n",
       "      <th>lfc_clone_pre_1_pvals_adj</th>\n",
       "      <th>salient_genes</th>\n",
       "      <th>salient_genes_rank</th>\n",
       "      <th>lfc_post_rank</th>\n",
       "      <th>lfc_pre_rank</th>\n",
       "      <th>lfc_clone_pre_10_rank</th>\n",
       "      <th>lfc_clone_pre_1_rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>0.841133</td>\n",
       "      <td>-0.794820</td>\n",
       "      <td>-0.725561</td>\n",
       "      <td>-1.166667</td>\n",
       "      <td>9.641343e-08</td>\n",
       "      <td>0.00015</td>\n",
       "      <td>1.121982e-12</td>\n",
       "      <td>2.611040e-41</td>\n",
       "      <td>-0.000007</td>\n",
       "      <td>10791.0</td>\n",
       "      <td>1865.0</td>\n",
       "      <td>17173.0</td>\n",
       "      <td>16486.0</td>\n",
       "      <td>17158.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>0.630395</td>\n",
       "      <td>-0.090250</td>\n",
       "      <td>-0.484623</td>\n",
       "      <td>0.291439</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>0.000031</td>\n",
       "      <td>2010.0</td>\n",
       "      <td>2357.0</td>\n",
       "      <td>13270.0</td>\n",
       "      <td>15601.0</td>\n",
       "      <td>6076.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A2M</td>\n",
       "      <td>-1.642448</td>\n",
       "      <td>2.252429</td>\n",
       "      <td>-0.178504</td>\n",
       "      <td>-0.586164</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>0.000051</td>\n",
       "      <td>658.0</td>\n",
       "      <td>18341.0</td>\n",
       "      <td>454.0</td>\n",
       "      <td>13850.0</td>\n",
       "      <td>13951.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A2M-AS1</td>\n",
       "      <td>0.501576</td>\n",
       "      <td>0.178894</td>\n",
       "      <td>0.893563</td>\n",
       "      <td>0.957310</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>-0.000027</td>\n",
       "      <td>15070.0</td>\n",
       "      <td>2735.0</td>\n",
       "      <td>8739.0</td>\n",
       "      <td>3887.0</td>\n",
       "      <td>2239.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A4GALT</td>\n",
       "      <td>-1.497684</td>\n",
       "      <td>2.251524</td>\n",
       "      <td>0.973174</td>\n",
       "      <td>-0.587068</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>6178.0</td>\n",
       "      <td>17228.0</td>\n",
       "      <td>1452.5</td>\n",
       "      <td>2731.5</td>\n",
       "      <td>14951.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19251</th>\n",
       "      <td>ZXDC</td>\n",
       "      <td>0.052903</td>\n",
       "      <td>-0.004820</td>\n",
       "      <td>0.134014</td>\n",
       "      <td>-0.018931</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>5125.0</td>\n",
       "      <td>6347.0</td>\n",
       "      <td>11984.0</td>\n",
       "      <td>9515.0</td>\n",
       "      <td>10352.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19252</th>\n",
       "      <td>ZYG11A</td>\n",
       "      <td>-1.497684</td>\n",
       "      <td>0.396684</td>\n",
       "      <td>2.828014</td>\n",
       "      <td>1.267772</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>-0.000029</td>\n",
       "      <td>15255.0</td>\n",
       "      <td>17228.0</td>\n",
       "      <td>6003.0</td>\n",
       "      <td>359.0</td>\n",
       "      <td>1522.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19253</th>\n",
       "      <td>ZYG11B</td>\n",
       "      <td>0.317693</td>\n",
       "      <td>-0.155747</td>\n",
       "      <td>-0.495282</td>\n",
       "      <td>-0.672724</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>-0.000004</td>\n",
       "      <td>10006.0</td>\n",
       "      <td>3636.0</td>\n",
       "      <td>14064.0</td>\n",
       "      <td>15651.0</td>\n",
       "      <td>16224.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19254</th>\n",
       "      <td>ZYX</td>\n",
       "      <td>-0.018861</td>\n",
       "      <td>-0.026237</td>\n",
       "      <td>0.016083</td>\n",
       "      <td>0.089319</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>-0.000028</td>\n",
       "      <td>15111.0</td>\n",
       "      <td>7555.0</td>\n",
       "      <td>12323.0</td>\n",
       "      <td>11443.0</td>\n",
       "      <td>8987.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19255</th>\n",
       "      <td>ZZEF1</td>\n",
       "      <td>0.036452</td>\n",
       "      <td>0.232192</td>\n",
       "      <td>0.355388</td>\n",
       "      <td>0.435197</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>0.000020</td>\n",
       "      <td>3480.0</td>\n",
       "      <td>6598.0</td>\n",
       "      <td>7955.0</td>\n",
       "      <td>6362.0</td>\n",
       "      <td>4521.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19256 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Gene  lfc_post   lfc_pre  lfc_clone_pre_10  lfc_clone_pre_1  \\\n",
       "0          A1BG  0.841133 -0.794820         -0.725561        -1.166667   \n",
       "1      A1BG-AS1  0.630395 -0.090250         -0.484623         0.291439   \n",
       "2           A2M -1.642448  2.252429         -0.178504        -0.586164   \n",
       "3       A2M-AS1  0.501576  0.178894          0.893563         0.957310   \n",
       "4        A4GALT -1.497684  2.251524          0.973174        -0.587068   \n",
       "...         ...       ...       ...               ...              ...   \n",
       "19251      ZXDC  0.052903 -0.004820          0.134014        -0.018931   \n",
       "19252    ZYG11A -1.497684  0.396684          2.828014         1.267772   \n",
       "19253    ZYG11B  0.317693 -0.155747         -0.495282        -0.672724   \n",
       "19254       ZYX -0.018861 -0.026237          0.016083         0.089319   \n",
       "19255     ZZEF1  0.036452  0.232192          0.355388         0.435197   \n",
       "\n",
       "       lfc_post_pvals_adj  lfc_pre_pvals_adj  lfc_clone_pre_10_pvals_adj  \\\n",
       "0            9.641343e-08            0.00015                1.121982e-12   \n",
       "1            1.000000e+00            1.00000                1.000000e+00   \n",
       "2            1.000000e+00            1.00000                1.000000e+00   \n",
       "3            1.000000e+00            1.00000                1.000000e+00   \n",
       "4                     NaN                NaN                         NaN   \n",
       "...                   ...                ...                         ...   \n",
       "19251        1.000000e+00            1.00000                1.000000e+00   \n",
       "19252                 NaN            1.00000                1.000000e+00   \n",
       "19253        1.000000e+00            1.00000                1.000000e+00   \n",
       "19254        1.000000e+00            1.00000                1.000000e+00   \n",
       "19255        1.000000e+00            1.00000                1.000000e+00   \n",
       "\n",
       "       lfc_clone_pre_1_pvals_adj  salient_genes  salient_genes_rank  \\\n",
       "0                   2.611040e-41      -0.000007             10791.0   \n",
       "1                   1.000000e+00       0.000031              2010.0   \n",
       "2                   1.000000e+00       0.000051               658.0   \n",
       "3                   1.000000e+00      -0.000027             15070.0   \n",
       "4                            NaN       0.000005              6178.0   \n",
       "...                          ...            ...                 ...   \n",
       "19251               1.000000e+00       0.000010              5125.0   \n",
       "19252               1.000000e+00      -0.000029             15255.0   \n",
       "19253               1.000000e+00      -0.000004             10006.0   \n",
       "19254               1.000000e+00      -0.000028             15111.0   \n",
       "19255               1.000000e+00       0.000020              3480.0   \n",
       "\n",
       "       lfc_post_rank  lfc_pre_rank  lfc_clone_pre_10_rank  \\\n",
       "0             1865.0       17173.0                16486.0   \n",
       "1             2357.0       13270.0                15601.0   \n",
       "2            18341.0         454.0                13850.0   \n",
       "3             2735.0        8739.0                 3887.0   \n",
       "4            17228.0        1452.5                 2731.5   \n",
       "...              ...           ...                    ...   \n",
       "19251         6347.0       11984.0                 9515.0   \n",
       "19252        17228.0        6003.0                  359.0   \n",
       "19253         3636.0       14064.0                15651.0   \n",
       "19254         7555.0       12323.0                11443.0   \n",
       "19255         6598.0        7955.0                 6362.0   \n",
       "\n",
       "       lfc_clone_pre_1_rank  \n",
       "0                   17158.0  \n",
       "1                    6076.0  \n",
       "2                   13951.0  \n",
       "3                    2239.0  \n",
       "4                   14951.5  \n",
       "...                     ...  \n",
       "19251               10352.0  \n",
       "19252                1522.0  \n",
       "19253               16224.0  \n",
       "19254                8987.0  \n",
       "19255                4521.0  \n",
       "\n",
       "[19256 rows x 15 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read in the whole dataset\n",
    "df = pd.read_csv(csv_path)\n",
    "df.rename(columns={'Unnamed: 0': 'Gene'}, inplace=True)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "29750"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read in the whole gene list\n",
    "with open('/home/che/TRIM/data_processed/combined_data_columns.npz', 'rb') as f:\n",
    "    npzfile = np.load(f, allow_pickle=True)\n",
    "    combined_data_columns = npzfile['cols']\n",
    "complete_gene_list = combined_data_columns.tolist()\n",
    "len(complete_gene_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "len top_genes_lfc_post 432\n",
      "len top_genes_lfc_pre 466\n",
      "len top_genes_grad: 386\n",
      "len top_genes_lfc_clone_pre_1 1519\n",
      "len top_genes_lfc_clone_pre_10 888\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>lfc_post</th>\n",
       "      <th>lfc_pre</th>\n",
       "      <th>lfc_clone_pre_10</th>\n",
       "      <th>lfc_clone_pre_1</th>\n",
       "      <th>lfc_post_pvals_adj</th>\n",
       "      <th>lfc_pre_pvals_adj</th>\n",
       "      <th>lfc_clone_pre_10_pvals_adj</th>\n",
       "      <th>lfc_clone_pre_1_pvals_adj</th>\n",
       "      <th>salient_genes</th>\n",
       "      <th>salient_genes_rank</th>\n",
       "      <th>lfc_post_rank</th>\n",
       "      <th>lfc_pre_rank</th>\n",
       "      <th>lfc_clone_pre_10_rank</th>\n",
       "      <th>lfc_clone_pre_1_rank</th>\n",
       "      <th>top_genes_lfc_post</th>\n",
       "      <th>top_genes_lfc_pre</th>\n",
       "      <th>top_genes_grad</th>\n",
       "      <th>top_genes_lfc_clone_pre_1</th>\n",
       "      <th>top_genes_lfc_clone_pre_10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>0.841133</td>\n",
       "      <td>-0.794820</td>\n",
       "      <td>-0.725561</td>\n",
       "      <td>-1.166667</td>\n",
       "      <td>9.641343e-08</td>\n",
       "      <td>1.504766e-04</td>\n",
       "      <td>1.121982e-12</td>\n",
       "      <td>2.611040e-41</td>\n",
       "      <td>-0.000007</td>\n",
       "      <td>10791.0</td>\n",
       "      <td>1865.0</td>\n",
       "      <td>17173.0</td>\n",
       "      <td>16486.0</td>\n",
       "      <td>17158.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>ABCB9</td>\n",
       "      <td>-0.504494</td>\n",
       "      <td>-1.068248</td>\n",
       "      <td>-0.942749</td>\n",
       "      <td>-1.228584</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>-0.000089</td>\n",
       "      <td>19153.0</td>\n",
       "      <td>14188.0</td>\n",
       "      <td>17735.0</td>\n",
       "      <td>17094.0</td>\n",
       "      <td>17257.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>ABCD2</td>\n",
       "      <td>0.612471</td>\n",
       "      <td>-0.986414</td>\n",
       "      <td>-0.716174</td>\n",
       "      <td>-0.660411</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>5.904538e-03</td>\n",
       "      <td>-0.000063</td>\n",
       "      <td>18558.0</td>\n",
       "      <td>2407.0</td>\n",
       "      <td>17579.0</td>\n",
       "      <td>16456.0</td>\n",
       "      <td>16187.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>ABHD17A</td>\n",
       "      <td>-0.496471</td>\n",
       "      <td>0.581175</td>\n",
       "      <td>0.784946</td>\n",
       "      <td>1.030809</td>\n",
       "      <td>2.574536e-03</td>\n",
       "      <td>9.659899e-08</td>\n",
       "      <td>3.985006e-25</td>\n",
       "      <td>1.507854e-35</td>\n",
       "      <td>0.000059</td>\n",
       "      <td>393.0</td>\n",
       "      <td>14141.0</td>\n",
       "      <td>4818.0</td>\n",
       "      <td>4126.0</td>\n",
       "      <td>2057.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>ABI3</td>\n",
       "      <td>-0.405581</td>\n",
       "      <td>0.551165</td>\n",
       "      <td>0.528436</td>\n",
       "      <td>1.170311</td>\n",
       "      <td>6.399717e-02</td>\n",
       "      <td>5.022182e-05</td>\n",
       "      <td>3.270137e-07</td>\n",
       "      <td>3.755932e-33</td>\n",
       "      <td>-0.000041</td>\n",
       "      <td>16932.0</td>\n",
       "      <td>13530.0</td>\n",
       "      <td>4982.0</td>\n",
       "      <td>5072.0</td>\n",
       "      <td>1738.0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19162</th>\n",
       "      <td>ZNF814</td>\n",
       "      <td>0.311633</td>\n",
       "      <td>-1.714411</td>\n",
       "      <td>-1.153668</td>\n",
       "      <td>-1.398973</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.870137e-05</td>\n",
       "      <td>-0.000070</td>\n",
       "      <td>18788.0</td>\n",
       "      <td>3671.0</td>\n",
       "      <td>18472.0</td>\n",
       "      <td>17487.0</td>\n",
       "      <td>17462.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19199</th>\n",
       "      <td>ZNHIT1</td>\n",
       "      <td>-0.080221</td>\n",
       "      <td>0.239010</td>\n",
       "      <td>0.256901</td>\n",
       "      <td>0.390924</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>6.796115e-03</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>4331.0</td>\n",
       "      <td>8756.0</td>\n",
       "      <td>7871.0</td>\n",
       "      <td>7497.0</td>\n",
       "      <td>4917.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19215</th>\n",
       "      <td>ZRANB2-AS1</td>\n",
       "      <td>-1.497684</td>\n",
       "      <td>4.351961</td>\n",
       "      <td>-1.127263</td>\n",
       "      <td>-2.687505</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>0.000075</td>\n",
       "      <td>158.0</td>\n",
       "      <td>17228.0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>17451.0</td>\n",
       "      <td>18660.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19222</th>\n",
       "      <td>ZSCAN18</td>\n",
       "      <td>0.466925</td>\n",
       "      <td>-0.539419</td>\n",
       "      <td>-1.463122</td>\n",
       "      <td>-1.888522</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>2.669569e-05</td>\n",
       "      <td>3.251631e-18</td>\n",
       "      <td>0.000020</td>\n",
       "      <td>3514.0</td>\n",
       "      <td>2869.0</td>\n",
       "      <td>16469.0</td>\n",
       "      <td>17919.0</td>\n",
       "      <td>17843.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19236</th>\n",
       "      <td>ZSWIM1</td>\n",
       "      <td>1.514728</td>\n",
       "      <td>-0.917021</td>\n",
       "      <td>-2.125969</td>\n",
       "      <td>-2.686929</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>6.870037e-03</td>\n",
       "      <td>-0.000019</td>\n",
       "      <td>13623.0</td>\n",
       "      <td>857.0</td>\n",
       "      <td>17451.0</td>\n",
       "      <td>18610.0</td>\n",
       "      <td>18656.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1925 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Gene  lfc_post   lfc_pre  lfc_clone_pre_10  lfc_clone_pre_1  \\\n",
       "0            A1BG  0.841133 -0.794820         -0.725561        -1.166667   \n",
       "37          ABCB9 -0.504494 -1.068248         -0.942749        -1.228584   \n",
       "45          ABCD2  0.612471 -0.986414         -0.716174        -0.660411   \n",
       "64        ABHD17A -0.496471  0.581175          0.784946         1.030809   \n",
       "76           ABI3 -0.405581  0.551165          0.528436         1.170311   \n",
       "...           ...       ...       ...               ...              ...   \n",
       "19162      ZNF814  0.311633 -1.714411         -1.153668        -1.398973   \n",
       "19199      ZNHIT1 -0.080221  0.239010          0.256901         0.390924   \n",
       "19215  ZRANB2-AS1 -1.497684  4.351961         -1.127263        -2.687505   \n",
       "19222     ZSCAN18  0.466925 -0.539419         -1.463122        -1.888522   \n",
       "19236      ZSWIM1  1.514728 -0.917021         -2.125969        -2.686929   \n",
       "\n",
       "       lfc_post_pvals_adj  lfc_pre_pvals_adj  lfc_clone_pre_10_pvals_adj  \\\n",
       "0            9.641343e-08       1.504766e-04                1.121982e-12   \n",
       "37           1.000000e+00       1.000000e+00                1.000000e+00   \n",
       "45           1.000000e+00       1.000000e+00                1.000000e+00   \n",
       "64           2.574536e-03       9.659899e-08                3.985006e-25   \n",
       "76           6.399717e-02       5.022182e-05                3.270137e-07   \n",
       "...                   ...                ...                         ...   \n",
       "19162        1.000000e+00       1.000000e+00                1.000000e+00   \n",
       "19199        1.000000e+00       1.000000e+00                1.000000e+00   \n",
       "19215                 NaN       1.000000e+00                1.000000e+00   \n",
       "19222        1.000000e+00       1.000000e+00                2.669569e-05   \n",
       "19236        1.000000e+00       1.000000e+00                1.000000e+00   \n",
       "\n",
       "       lfc_clone_pre_1_pvals_adj  salient_genes  salient_genes_rank  \\\n",
       "0                   2.611040e-41      -0.000007             10791.0   \n",
       "37                  1.000000e+00      -0.000089             19153.0   \n",
       "45                  5.904538e-03      -0.000063             18558.0   \n",
       "64                  1.507854e-35       0.000059               393.0   \n",
       "76                  3.755932e-33      -0.000041             16932.0   \n",
       "...                          ...            ...                 ...   \n",
       "19162               1.870137e-05      -0.000070             18788.0   \n",
       "19199               6.796115e-03       0.000015              4331.0   \n",
       "19215               1.000000e+00       0.000075               158.0   \n",
       "19222               3.251631e-18       0.000020              3514.0   \n",
       "19236               6.870037e-03      -0.000019             13623.0   \n",
       "\n",
       "       lfc_post_rank  lfc_pre_rank  lfc_clone_pre_10_rank  \\\n",
       "0             1865.0       17173.0                16486.0   \n",
       "37           14188.0       17735.0                17094.0   \n",
       "45            2407.0       17579.0                16456.0   \n",
       "64           14141.0        4818.0                 4126.0   \n",
       "76           13530.0        4982.0                 5072.0   \n",
       "...              ...           ...                    ...   \n",
       "19162         3671.0       18472.0                17487.0   \n",
       "19199         8756.0        7871.0                 7497.0   \n",
       "19215        17228.0          51.0                17451.0   \n",
       "19222         2869.0       16469.0                17919.0   \n",
       "19236          857.0       17451.0                18610.0   \n",
       "\n",
       "       lfc_clone_pre_1_rank  top_genes_lfc_post  top_genes_lfc_pre  \\\n",
       "0                   17158.0                True               True   \n",
       "37                  17257.0               False              False   \n",
       "45                  16187.0               False              False   \n",
       "64                   2057.0                True               True   \n",
       "76                   1738.0               False               True   \n",
       "...                     ...                 ...                ...   \n",
       "19162               17462.0               False              False   \n",
       "19199                4917.0               False              False   \n",
       "19215               18660.0               False              False   \n",
       "19222               17843.0               False              False   \n",
       "19236               18656.0               False              False   \n",
       "\n",
       "       top_genes_grad  top_genes_lfc_clone_pre_1  top_genes_lfc_clone_pre_10  \n",
       "0               False                       True                        True  \n",
       "37               True                      False                       False  \n",
       "45              False                       True                       False  \n",
       "64              False                       True                        True  \n",
       "76              False                       True                        True  \n",
       "...               ...                        ...                         ...  \n",
       "19162           False                       True                       False  \n",
       "19199           False                       True                       False  \n",
       "19215            True                      False                       False  \n",
       "19222           False                       True                        True  \n",
       "19236           False                       True                       False  \n",
       "\n",
       "[1925 rows x 20 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# first subset to the genes found by different methods\n",
    "if de_type == 'no_emergent_no_disappear':\n",
    "    alpha = 0.05\n",
    "    log2fc_threshold = -np.inf\n",
    "    high_gradient_threshold = 0.996\n",
    "    low_gradient_threshold = 1- high_gradient_threshold #0.01\n",
    "else:\n",
    "    alpha = 0.01\n",
    "    log2fc_threshold = -np.inf\n",
    "    high_gradient_threshold = 0.99\n",
    "    low_gradient_threshold = 1- high_gradient_threshold #0.01\n",
    "\n",
    "top_genes_lfc_post = df[df['lfc_post_pvals_adj'] < alpha].copy()\n",
    "top_genes_lfc_post = top_genes_lfc_post[abs(top_genes_lfc_post['lfc_post']) > log2fc_threshold]\n",
    "df['top_genes_lfc_post'] = df['Gene'].isin(top_genes_lfc_post['Gene'])\n",
    "print('len top_genes_lfc_post', len(top_genes_lfc_post))\n",
    "\n",
    "top_genes_lfc_pre = df[df['lfc_pre_pvals_adj'] < alpha].copy()\n",
    "top_genes_lfc_pre = top_genes_lfc_pre[abs(top_genes_lfc_pre['lfc_pre']) > log2fc_threshold]\n",
    "df['top_genes_lfc_pre'] = df['Gene'].isin(top_genes_lfc_pre['Gene'])\n",
    "print('len top_genes_lfc_pre', len(top_genes_lfc_pre))\n",
    "\n",
    "top_genes_grad = df[(df['salient_genes'] > df['salient_genes'].quantile(high_gradient_threshold)) | \n",
    "                    (df['salient_genes'] < df['salient_genes'].quantile(low_gradient_threshold))]\n",
    "df['top_genes_grad'] = df['Gene'].isin(top_genes_grad['Gene'])\n",
    "print('len top_genes_grad:', len(top_genes_grad))\n",
    "\n",
    "top_genes_lfc_clone_pre_1 = df[df['lfc_clone_pre_1_pvals_adj'] < alpha]\n",
    "top_genes_lfc_clone_pre_1 = top_genes_lfc_clone_pre_1[abs(top_genes_lfc_clone_pre_1['lfc_clone_pre_1']) > log2fc_threshold]\n",
    "df['top_genes_lfc_clone_pre_1'] = df['Gene'].isin(top_genes_lfc_clone_pre_1['Gene'])\n",
    "print('len top_genes_lfc_clone_pre_1', len(top_genes_lfc_clone_pre_1))\n",
    "\n",
    "top_genes_lfc_clone_pre_10 = df[df['lfc_clone_pre_10_pvals_adj'] < alpha]\n",
    "top_genes_lfc_clone_pre_10 = top_genes_lfc_clone_pre_10[abs(top_genes_lfc_clone_pre_10['lfc_clone_pre_10']) > log2fc_threshold]\n",
    "df['top_genes_lfc_clone_pre_10'] = df['Gene'].isin(top_genes_lfc_clone_pre_10['Gene'])\n",
    "print('len top_genes_lfc_clone_pre_10', len(top_genes_lfc_clone_pre_10))\n",
    "\n",
    "df_subset = df[df['Gene'].isin(top_genes_lfc_post['Gene']) | \n",
    "                df['Gene'].isin(top_genes_lfc_pre['Gene']) |\n",
    "               df['Gene'].isin(top_genes_grad['Gene']) | \n",
    "               df['Gene'].isin(top_genes_lfc_clone_pre_10['Gene']) | \n",
    "               df['Gene'].isin(top_genes_lfc_clone_pre_1['Gene'])].copy()\n",
    "df_subset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "len protein_coding_genes: 19991\n"
     ]
    }
   ],
   "source": [
    "# read in gencode file\n",
    "protein_gene_path = '/home/che/TRIM/git/tcr/other_information/gencode.v41.annotation.gtf'\n",
    "gr = pr.read_gtf(protein_gene_path)\n",
    "gencode_df = gr.df\n",
    "gencode_df = gencode_df[gencode_df['Feature'] == 'gene']\n",
    "gencode_df = gencode_df[gencode_df['gene_type'] == 'protein_coding']\n",
    "protein_coding_genes = gencode_df['gene_name'].unique().tolist()\n",
    "print('len protein_coding_genes:', len(protein_coding_genes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_720346/2460345661.py:3: DtypeWarning: Columns (35,38,45,48,49) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  protein_genes = pd.read_csv(protein_gene_path, sep='\\t')\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "len protein_genes_2: 19268\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "987"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# protein coding genes downloaded from: https://www.genenames.org/download/statistics-and-files/\n",
    "protein_gene_path = '/home/che/TRIM/git/tcr/other_information/gene_with_protein_product.txt'\n",
    "protein_genes = pd.read_csv(protein_gene_path, sep='\\t')\n",
    "protein_genes_2 = protein_genes['symbol'].unique().tolist()\n",
    "print('len protein_genes_2:', len(protein_genes_2))\n",
    "len(set(protein_coding_genes) - set(protein_genes_2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>lfc_post</th>\n",
       "      <th>lfc_pre</th>\n",
       "      <th>lfc_clone_pre_10</th>\n",
       "      <th>lfc_clone_pre_1</th>\n",
       "      <th>lfc_post_pvals_adj</th>\n",
       "      <th>lfc_pre_pvals_adj</th>\n",
       "      <th>lfc_clone_pre_10_pvals_adj</th>\n",
       "      <th>lfc_clone_pre_1_pvals_adj</th>\n",
       "      <th>salient_genes</th>\n",
       "      <th>salient_genes_rank</th>\n",
       "      <th>lfc_post_rank</th>\n",
       "      <th>lfc_pre_rank</th>\n",
       "      <th>lfc_clone_pre_10_rank</th>\n",
       "      <th>lfc_clone_pre_1_rank</th>\n",
       "      <th>top_genes_lfc_post</th>\n",
       "      <th>top_genes_lfc_pre</th>\n",
       "      <th>top_genes_grad</th>\n",
       "      <th>top_genes_lfc_clone_pre_1</th>\n",
       "      <th>top_genes_lfc_clone_pre_10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>0.841133</td>\n",
       "      <td>-0.794820</td>\n",
       "      <td>-0.725561</td>\n",
       "      <td>-1.166667</td>\n",
       "      <td>9.641343e-08</td>\n",
       "      <td>1.504766e-04</td>\n",
       "      <td>1.121982e-12</td>\n",
       "      <td>2.611040e-41</td>\n",
       "      <td>-0.000007</td>\n",
       "      <td>10791.0</td>\n",
       "      <td>1865.0</td>\n",
       "      <td>17173.0</td>\n",
       "      <td>16486.0</td>\n",
       "      <td>17158.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>ABCB9</td>\n",
       "      <td>-0.504494</td>\n",
       "      <td>-1.068248</td>\n",
       "      <td>-0.942749</td>\n",
       "      <td>-1.228584</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>-0.000089</td>\n",
       "      <td>19153.0</td>\n",
       "      <td>14188.0</td>\n",
       "      <td>17735.0</td>\n",
       "      <td>17094.0</td>\n",
       "      <td>17257.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>ABCD2</td>\n",
       "      <td>0.612471</td>\n",
       "      <td>-0.986414</td>\n",
       "      <td>-0.716174</td>\n",
       "      <td>-0.660411</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>5.904538e-03</td>\n",
       "      <td>-0.000063</td>\n",
       "      <td>18558.0</td>\n",
       "      <td>2407.0</td>\n",
       "      <td>17579.0</td>\n",
       "      <td>16456.0</td>\n",
       "      <td>16187.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>ABHD17A</td>\n",
       "      <td>-0.496471</td>\n",
       "      <td>0.581175</td>\n",
       "      <td>0.784946</td>\n",
       "      <td>1.030809</td>\n",
       "      <td>2.574536e-03</td>\n",
       "      <td>9.659899e-08</td>\n",
       "      <td>3.985006e-25</td>\n",
       "      <td>1.507854e-35</td>\n",
       "      <td>0.000059</td>\n",
       "      <td>393.0</td>\n",
       "      <td>14141.0</td>\n",
       "      <td>4818.0</td>\n",
       "      <td>4126.0</td>\n",
       "      <td>2057.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>ABI3</td>\n",
       "      <td>-0.405581</td>\n",
       "      <td>0.551165</td>\n",
       "      <td>0.528436</td>\n",
       "      <td>1.170311</td>\n",
       "      <td>6.399717e-02</td>\n",
       "      <td>5.022182e-05</td>\n",
       "      <td>3.270137e-07</td>\n",
       "      <td>3.755932e-33</td>\n",
       "      <td>-0.000041</td>\n",
       "      <td>16932.0</td>\n",
       "      <td>13530.0</td>\n",
       "      <td>4982.0</td>\n",
       "      <td>5072.0</td>\n",
       "      <td>1738.0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19134</th>\n",
       "      <td>ZNF776</td>\n",
       "      <td>0.372331</td>\n",
       "      <td>-1.575151</td>\n",
       "      <td>-1.176315</td>\n",
       "      <td>-1.623787</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>2.243293e-03</td>\n",
       "      <td>-0.000053</td>\n",
       "      <td>18020.0</td>\n",
       "      <td>3317.0</td>\n",
       "      <td>18361.0</td>\n",
       "      <td>17529.0</td>\n",
       "      <td>17655.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19162</th>\n",
       "      <td>ZNF814</td>\n",
       "      <td>0.311633</td>\n",
       "      <td>-1.714411</td>\n",
       "      <td>-1.153668</td>\n",
       "      <td>-1.398973</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.870137e-05</td>\n",
       "      <td>-0.000070</td>\n",
       "      <td>18788.0</td>\n",
       "      <td>3671.0</td>\n",
       "      <td>18472.0</td>\n",
       "      <td>17487.0</td>\n",
       "      <td>17462.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19199</th>\n",
       "      <td>ZNHIT1</td>\n",
       "      <td>-0.080221</td>\n",
       "      <td>0.239010</td>\n",
       "      <td>0.256901</td>\n",
       "      <td>0.390924</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>6.796115e-03</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>4331.0</td>\n",
       "      <td>8756.0</td>\n",
       "      <td>7871.0</td>\n",
       "      <td>7497.0</td>\n",
       "      <td>4917.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19222</th>\n",
       "      <td>ZSCAN18</td>\n",
       "      <td>0.466925</td>\n",
       "      <td>-0.539419</td>\n",
       "      <td>-1.463122</td>\n",
       "      <td>-1.888522</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>2.669569e-05</td>\n",
       "      <td>3.251631e-18</td>\n",
       "      <td>0.000020</td>\n",
       "      <td>3514.0</td>\n",
       "      <td>2869.0</td>\n",
       "      <td>16469.0</td>\n",
       "      <td>17919.0</td>\n",
       "      <td>17843.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19236</th>\n",
       "      <td>ZSWIM1</td>\n",
       "      <td>1.514728</td>\n",
       "      <td>-0.917021</td>\n",
       "      <td>-2.125969</td>\n",
       "      <td>-2.686929</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>6.870037e-03</td>\n",
       "      <td>-0.000019</td>\n",
       "      <td>13623.0</td>\n",
       "      <td>857.0</td>\n",
       "      <td>17451.0</td>\n",
       "      <td>18610.0</td>\n",
       "      <td>18656.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1628 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          Gene  lfc_post   lfc_pre  lfc_clone_pre_10  lfc_clone_pre_1  \\\n",
       "0         A1BG  0.841133 -0.794820         -0.725561        -1.166667   \n",
       "37       ABCB9 -0.504494 -1.068248         -0.942749        -1.228584   \n",
       "45       ABCD2  0.612471 -0.986414         -0.716174        -0.660411   \n",
       "64     ABHD17A -0.496471  0.581175          0.784946         1.030809   \n",
       "76        ABI3 -0.405581  0.551165          0.528436         1.170311   \n",
       "...        ...       ...       ...               ...              ...   \n",
       "19134   ZNF776  0.372331 -1.575151         -1.176315        -1.623787   \n",
       "19162   ZNF814  0.311633 -1.714411         -1.153668        -1.398973   \n",
       "19199   ZNHIT1 -0.080221  0.239010          0.256901         0.390924   \n",
       "19222  ZSCAN18  0.466925 -0.539419         -1.463122        -1.888522   \n",
       "19236   ZSWIM1  1.514728 -0.917021         -2.125969        -2.686929   \n",
       "\n",
       "       lfc_post_pvals_adj  lfc_pre_pvals_adj  lfc_clone_pre_10_pvals_adj  \\\n",
       "0            9.641343e-08       1.504766e-04                1.121982e-12   \n",
       "37           1.000000e+00       1.000000e+00                1.000000e+00   \n",
       "45           1.000000e+00       1.000000e+00                1.000000e+00   \n",
       "64           2.574536e-03       9.659899e-08                3.985006e-25   \n",
       "76           6.399717e-02       5.022182e-05                3.270137e-07   \n",
       "...                   ...                ...                         ...   \n",
       "19134        1.000000e+00       1.000000e+00                1.000000e+00   \n",
       "19162        1.000000e+00       1.000000e+00                1.000000e+00   \n",
       "19199        1.000000e+00       1.000000e+00                1.000000e+00   \n",
       "19222        1.000000e+00       1.000000e+00                2.669569e-05   \n",
       "19236        1.000000e+00       1.000000e+00                1.000000e+00   \n",
       "\n",
       "       lfc_clone_pre_1_pvals_adj  salient_genes  salient_genes_rank  \\\n",
       "0                   2.611040e-41      -0.000007             10791.0   \n",
       "37                  1.000000e+00      -0.000089             19153.0   \n",
       "45                  5.904538e-03      -0.000063             18558.0   \n",
       "64                  1.507854e-35       0.000059               393.0   \n",
       "76                  3.755932e-33      -0.000041             16932.0   \n",
       "...                          ...            ...                 ...   \n",
       "19134               2.243293e-03      -0.000053             18020.0   \n",
       "19162               1.870137e-05      -0.000070             18788.0   \n",
       "19199               6.796115e-03       0.000015              4331.0   \n",
       "19222               3.251631e-18       0.000020              3514.0   \n",
       "19236               6.870037e-03      -0.000019             13623.0   \n",
       "\n",
       "       lfc_post_rank  lfc_pre_rank  lfc_clone_pre_10_rank  \\\n",
       "0             1865.0       17173.0                16486.0   \n",
       "37           14188.0       17735.0                17094.0   \n",
       "45            2407.0       17579.0                16456.0   \n",
       "64           14141.0        4818.0                 4126.0   \n",
       "76           13530.0        4982.0                 5072.0   \n",
       "...              ...           ...                    ...   \n",
       "19134         3317.0       18361.0                17529.0   \n",
       "19162         3671.0       18472.0                17487.0   \n",
       "19199         8756.0        7871.0                 7497.0   \n",
       "19222         2869.0       16469.0                17919.0   \n",
       "19236          857.0       17451.0                18610.0   \n",
       "\n",
       "       lfc_clone_pre_1_rank  top_genes_lfc_post  top_genes_lfc_pre  \\\n",
       "0                   17158.0                True               True   \n",
       "37                  17257.0               False              False   \n",
       "45                  16187.0               False              False   \n",
       "64                   2057.0                True               True   \n",
       "76                   1738.0               False               True   \n",
       "...                     ...                 ...                ...   \n",
       "19134               17655.0               False              False   \n",
       "19162               17462.0               False              False   \n",
       "19199                4917.0               False              False   \n",
       "19222               17843.0               False              False   \n",
       "19236               18656.0               False              False   \n",
       "\n",
       "       top_genes_grad  top_genes_lfc_clone_pre_1  top_genes_lfc_clone_pre_10  \n",
       "0               False                       True                        True  \n",
       "37               True                      False                       False  \n",
       "45              False                       True                       False  \n",
       "64              False                       True                        True  \n",
       "76              False                       True                        True  \n",
       "...               ...                        ...                         ...  \n",
       "19134           False                       True                       False  \n",
       "19162           False                       True                       False  \n",
       "19199           False                       True                       False  \n",
       "19222           False                       True                        True  \n",
       "19236           False                       True                       False  \n",
       "\n",
       "[1628 rows x 20 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Subset to human protein-coding genes\n",
    "df_subset = df_subset[df_subset['Gene'].isin(protein_coding_genes)]\n",
    "df_subset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(401, 431, 808, 1408, 219)"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "top_genes_lfc_post = set(df_subset[df_subset['top_genes_lfc_post'] == True]['Gene'].tolist())\n",
    "top_genes_lfc_pre = set(df_subset[df_subset['top_genes_lfc_pre'] == True]['Gene'].tolist())\n",
    "top_genes_grad = set(df_subset[df_subset['top_genes_grad'] == True]['Gene'].tolist())\n",
    "top_genes_lfc_clone_pre_10 = set(df_subset[df_subset['top_genes_lfc_clone_pre_10'] == True]['Gene'].tolist())\n",
    "top_genes_lfc_clone_pre_1 = set(df_subset[df_subset['top_genes_lfc_clone_pre_1'] == True]['Gene'].tolist())\n",
    "\n",
    "len(top_genes_lfc_post), len(top_genes_lfc_pre), \\\n",
    "        len(top_genes_lfc_clone_pre_10), len(top_genes_lfc_clone_pre_1), \\\n",
    "                len(top_genes_grad)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Now, perform gene set enrichment analysis for each list\n",
    "\n",
    "- gene_list: important genes found by each method\n",
    "- backgroun: all genes in the genome (measured in RNA-seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Helper functions to remove some gene sets with high overlap --------\n",
    "from itertools import combinations\n",
    "def jaccard_similarity(set1, set2):\n",
    "    \"\"\"Calculate Jaccard similarity between two sets\"\"\"\n",
    "    intersection = len(set1.intersection(set2))\n",
    "    union = len(set1.union(set2))\n",
    "    return intersection / union if union > 0 else 0\n",
    "\n",
    "def filter_enrichr_by_jaccard_similarity(enr, threshold=0.5, column_name='Genes', \n",
    "                                         sort_by='Adjusted P-value', ascending=True):\n",
    "    \"\"\"\n",
    "    Filter Enrichr results to keep only gene sets with minimal overlap\n",
    "    \n",
    "    Parameters:\n",
    "    -----------\n",
    "    enr : pandas.DataFrame\n",
    "        Enrichr results dataframe containing gene sets\n",
    "    threshold : float, default=0.5\n",
    "        Maximum allowed Jaccard similarity between kept gene sets\n",
    "    column_name : str, default='Genes'\n",
    "        Column name containing the genes (typically semicolon or comma separated)\n",
    "    sort_by : str, default='Adjusted P-value'\n",
    "        Column to sort by before filtering\n",
    "    ascending : bool, default=True\n",
    "        Whether to sort in ascending order\n",
    "        \n",
    "    Returns:\n",
    "    --------\n",
    "    pandas.DataFrame\n",
    "        Filtered Enrichr results with minimal gene set overlap\n",
    "    \"\"\"\n",
    "    # Make a copy to avoid modifying the original\n",
    "    df = enr.copy()\n",
    "    \n",
    "    # Convert gene lists to sets\n",
    "    gene_sets = []\n",
    "    for gene_list in df[column_name]:\n",
    "        # Handle different separators (comma, semicolon, etc.)\n",
    "        if isinstance(gene_list, str):\n",
    "            if ';' in gene_list:\n",
    "                genes = set(gene_list.split(';'))\n",
    "            elif ',' in gene_list:\n",
    "                genes = set(gene_list.split(','))\n",
    "            else:\n",
    "                genes = set(gene_list.split())\n",
    "            gene_sets.append(genes)\n",
    "        else:\n",
    "            gene_sets.append(set())\n",
    "    \n",
    "    # Sort by specified column (usually p-value or adjusted p-value)\n",
    "    df['gene_sets'] = gene_sets\n",
    "    df = df.sort_values(by=sort_by, ascending=ascending).reset_index(drop=True)\n",
    "    \n",
    "    # Initialize list to keep track of which rows to keep\n",
    "    to_keep = [True] * len(df)\n",
    "    \n",
    "    # Compare each gene set with all preceding gene sets that we've decided to keep\n",
    "    for i in range(1, len(df)):\n",
    "        if not to_keep[i]:  # Skip if already marked for removal\n",
    "            continue\n",
    "            \n",
    "        for j in range(i):\n",
    "            if not to_keep[j]:  # Skip if already marked for removal\n",
    "                continue\n",
    "                \n",
    "            # Calculate Jaccard similarity\n",
    "            similarity = jaccard_similarity(df.loc[i, 'gene_sets'], df.loc[j, 'gene_sets'])\n",
    "            \n",
    "            # If similarity exceeds threshold, mark the current gene set for removal\n",
    "            if similarity > threshold:\n",
    "                to_keep[i] = False\n",
    "                break\n",
    "            # else:\n",
    "            #     print(f\"Gene set {i} and {j} have Jaccard similarity: {similarity:.2f}\")\n",
    "    \n",
    "    # Filter the dataframe\n",
    "    filtered_df = df[to_keep].copy()\n",
    "    \n",
    "    # Remove the temporary gene_sets column\n",
    "    filtered_df.drop(columns=['gene_sets'], inplace=True)\n",
    "    \n",
    "    return filtered_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "methods_list = [top_genes_grad, top_genes_lfc_post, top_genes_lfc_pre, top_genes_lfc_clone_pre_1, top_genes_lfc_clone_pre_10]\n",
    "methods_name = ['top_genes_grad', 'top_genes_lfc_post', 'top_genes_lfc_pre', 'top_genes_lfc_clone_pre_1', 'top_genes_lfc_clone_pre_10']\n",
    "geneset_enrich_list = []\n",
    "\n",
    "for top_genes_temp in methods_list:\n",
    "    gene_list_temp = list(top_genes_temp)\n",
    "    background_genes = complete_gene_list\n",
    "\n",
    "    # Run enrichment analysis\n",
    "    enr = gp.enrichr(\n",
    "            gene_list=gene_list_temp,\n",
    "            gene_sets='GO_Biological_Process_2018', \n",
    "            organism='Human',\n",
    "            background=background_genes,\n",
    "            outdir=None,\n",
    "            cutoff=0.05,\n",
    "        )\n",
    "    enr_df_temp = enr.results\n",
    "    enr_df_temp.to_csv(f'{geneset_path}/{methods_name[methods_list.index(top_genes_temp)]}_enrichment.csv', index=False)\n",
    "    enr_df_temp = enr_df_temp[enr_df_temp['Adjusted P-value'] < 0.1]\n",
    "    \n",
    "    filtered_enr = filter_enrichr_by_jaccard_similarity(enr_df_temp, threshold=0.5)\n",
    "    # save filtered_enr in geneset_path\n",
    "    filtered_enr.to_csv(f'{geneset_path}/{methods_name[methods_list.index(top_genes_temp)]}_filtered_enrichment.csv', index=False)\n",
    "    # get the top 15 list of genesets with adjusted p-value < 0.1\n",
    "    geneset_enrich_temp = filtered_enr['Term'].tolist()[:15]\n",
    "    geneset_enrich_list.append(geneset_enrich_temp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "# flatten geneset_enrich_list\n",
    "geneset_enrich_list_flat = [item for sublist in geneset_enrich_list for item in sublist]\n",
    "geneset_enrich_list_flat = list(set(geneset_enrich_list_flat))\n",
    "len(geneset_enrich_list_flat)\n",
    "# save geneset_enrich_list_flat in geneset_path\n",
    "with open(f'{geneset_path}/geneset_enrich_list.json', 'w') as f:\n",
    "    json.dump(geneset_enrich_list_flat, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['positive regulation of cell proliferation (GO:0008284)',\n",
       " 'nuclear-transcribed mRNA catabolic process, nonsense-mediated decay (GO:0000184)',\n",
       " 'SRP-dependent cotranslational protein targeting to membrane (GO:0006614)',\n",
       " 'positive regulation of peptidyl-tyrosine phosphorylation (GO:0050731)',\n",
       " 'platelet degranulation (GO:0002576)',\n",
       " 'T cell receptor signaling pathway (GO:0050852)',\n",
       " 'T cell activation (GO:0042110)',\n",
       " 'ribosomal small subunit biogenesis (GO:0042274)',\n",
       " 'regulation of apoptotic process (GO:0042981)',\n",
       " 'regulation of small GTPase mediated signal transduction (GO:0051056)',\n",
       " 'glomerular epithelial cell differentiation (GO:0072311)',\n",
       " 'immunological synapse formation (GO:0001771)',\n",
       " 'regulation of cell proliferation (GO:0042127)',\n",
       " 'ribosomal large subunit biogenesis (GO:0042273)',\n",
       " 'regulation of immune response (GO:0050776)',\n",
       " 'ribosome assembly (GO:0042255)',\n",
       " 'neutrophil degranulation (GO:0043312)',\n",
       " 'cellular response to interferon-gamma (GO:0071346)',\n",
       " 'cytokine-mediated signaling pathway (GO:0019221)',\n",
       " 'positive regulation of apoptotic process (GO:0043065)',\n",
       " 'negative regulation of cellular process (GO:0048523)',\n",
       " 'cellular response to DNA damage stimulus (GO:0006974)',\n",
       " 'transmembrane receptor protein tyrosine kinase signaling pathway (GO:0007169)',\n",
       " 'positive regulation of cellular process (GO:0048522)',\n",
       " 'negative regulation of apoptotic process (GO:0043066)',\n",
       " 'inflammatory response (GO:0006954)',\n",
       " 'cellular response to cytokine stimulus (GO:0071345)',\n",
       " 'positive regulation of lipid kinase activity (GO:0090218)',\n",
       " 'coronary vasculature morphogenesis (GO:0060977)',\n",
       " 'positive regulation of cytosolic calcium ion concentration (GO:0007204)',\n",
       " 'cellular defense response (GO:0006968)',\n",
       " 'positive regulation of intracellular signal transduction (GO:1902533)',\n",
       " 'antigen processing and presentation of exogenous peptide antigen via MHC class I (GO:0042590)',\n",
       " 'apoptotic process (GO:0006915)',\n",
       " 'regulation of I-kappaB kinase/NF-kappaB signaling (GO:0043122)',\n",
       " 'negative regulation of cell proliferation (GO:0008285)',\n",
       " 'neutrophil activation involved in immune response (GO:0002283)',\n",
       " 'positive regulation of glutamate receptor signaling pathway (GO:1900451)',\n",
       " 'positive regulation of MAP kinase activity (GO:0043406)',\n",
       " 'cytoplasmic translation (GO:0002181)']"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "geneset_enrich_list_flat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Download the GO:BP library from Enrichr (as a dictionary)\n",
    "go_bp_dict = gp.get_library(name='GO_Biological_Process_2018', organism='Human')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "go_enrichr_all = {}\n",
    "for go in geneset_enrich_list_flat:\n",
    "    go_enrichr_all[go] = go_bp_dict[go]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(f'{geneset_path}/go_enrichr.json', 'w') as f:\n",
    "    json.dump(go_enrichr_all, f, indent=4)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "tcr2",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
